packages/store-sync/src/recs/syncToRecs.ts

```diff
@@ -0,0 +1,227 @@
+import { StoreConfig, storeEventsAbi } from "@latticexyz/store";
+import { Address, Block, Chain, Hex, PublicClient, TransactionReceipt, Transport } from "viem";
+import {
+  ComponentValue,
+  Entity,
+  Component as RecsComponent,
+  Schema as RecsSchema,
+  World as RecsWorld,
+  getComponentValue,
+  setComponent,
+} from "@latticexyz/recs";
+import { BlockLogs, Table, TableRecord } from "../common";
+import {
+  createBlockStream,
+  isNonPendingBlock,
+  blockRangeToLogs,
+  groupLogsByBlockNumber,
+} from "@latticexyz/block-logs-stream";
+import { BehaviorSubject, filter, map, tap, mergeMap, from, concatMap, Observable, share, firstValueFrom } from "rxjs";
+import { BlockStorageOperations, blockLogsToStorage } from "../blockLogsToStorage";
+import { recsStorage } from "./recsStorage";
+import { hexKeyTupleToEntity } from "./hexKeyTupleToEntity";
+import { debug } from "./debug";
+import { defineInternalComponents } from "./defineInternalComponents";
+import { getTableKey } from "./getTableKey";
+import { StoreComponentMetadata, SyncStep } from "./common";
+import { encodeEntity } from "./encodeEntity";
+
+type SyncToRecsOptions<
+  TConfig extends StoreConfig = StoreConfig,
+  TComponents extends Record<string, RecsComponent<RecsSchema, StoreComponentMetadata>> = Record<
+    string,
+    RecsComponent<RecsSchema, StoreComponentMetadata>
+  >
+> = {
+  world: RecsWorld;
+  config: TConfig;
+  address: Address;
+  // TODO: make this optional and return one if none provided (but will need chain ID at least)
+  publicClient: PublicClient<Transport, Chain>;
+  // TODO: generate these from config and return instead?
+  components: TComponents;
+  initialState?: {
+    blockNumber: bigint | null;
+    tables: (Table & { records: TableRecord[] })[];
+  };
```

Comment:
I originally built this using `indexerUrl` that would fetch from the indexer here, but it caused a circular dependency (because of tRPC types).

@alvrs and I talked about moving the indexer code into `store-sync` package and reducing the `store-indexer` package to just a "indexer runner", but now that I am looking back at this, I quite like the flexibility of being able to provide the initial state from ~anywhere, rather than strictly our indexer.

Maybe a good middle ground is exposing a helper in `store-indexer` that is `getInitialState(indexerUrl, chainId, address)` that abstracts over the raw tRPC call:

<img width="511" alt="image" src="https://github.com/latticexyz/mud/assets/508855/ea11f4cd-5879-4fd4-a436-93000f80b03b">

But honestly, this isn't much code, so maybe it's fine how it is?

Reply:
>  I quite like the flexibility of being able to provide the initial state from ~anywhere, rather than strictly our indexer.

agree

> It would be nice to remove the need to know about our transformer, AppRouter type, etc. so maybe an abstraction layer over it is nice.

agree

What if we move it to one package but keep it modular (so users could do it as shown in your example) but also have a lean wrapper around it with simple options for the "default setup"?

Reply:
okay, took another pass and got it down to an option in `syncToRecs`

<img width="443" alt="image" src="https://github.com/latticexyz/mud/assets/508855/475927a0-7c84-4017-a061-1a2723d77ce3">

internally in `syncToRecs`, it does this:

<img width="601" alt="image" src="https://github.com/latticexyz/mud/assets/508855/30fddd1a-0179-4a1c-be29-70622f5b7084">

Reply:
updated store-indexer package branch so it's built on top of this one: https://github.com/latticexyz/mud/pull/1198